python oop principles module - closes #675 (#677)
bob-crutchley/courseware: topics/python/modules/object-oriented-programming/README.md

![class](https://qa-courseware-images.s3.eu-west-2.amazonaws.com/python/object-oriented-programming/000.png)

When a child class inherits from a parents class, and there is no default constructors.
The first thing the child class should do is run the parent constructor. Say we have a bird class and a child class of an owl.

```python
class bird:
    def __init__(self, fly)
    self.fly = True

class Owl(Bird):
    def __init__(self, fly, babies):
        Bird.__init__(fly)
        self.babies = babies

    def reproduce(self):
        self.babies += 1
```

### Polymorphism

Polymorphism means many forms.
The reason we use `inheritance` is so a subclass can use the `data members` from a parent class.
However different child classes may act `differently` in regards to a data member inherited from the superclass.
This is polymorphism.
So if we consider our animal superclass, with cat and dog subclasses. We can have a method called `speak`. A cat's response to the method would be `meow` and the dogs class response would be `woof`.
There are two types of polymorphism, overloading and overriding.

> 

Reviewer: Need to a comment to say that this is overloading.

Author: ```suggestion
Overloading is the action of defining multiple methods with the same name. This is an example of overloading.
```

```python
class Bird:
    def __init__(self, wingspan):
        self.wingspan = wingspan

Eagle = Bird(104)

len(Eagle)
```

> This will produce an error as the `len` method has no logic has been assign to it when a Bird type is assign to it.

Reviewer: This does not read very well.

> This will produce an error as the `len` method has no logic has been assign to it when a Bird type is assign to it.

Author: ```suggestion
This will produce an error as the `len` method has no logic assigned to it in the `Bird` class.
```
